Implement `GetAccountState` operation for Jörmungandr API Client

jonathanknowles opened the pull request:
# Issue Number

#902 

# Overview

This PR:

- [x] Adds an `AccountId` type with `ToText`/`FromText` instances, along with serialization tests.
- [x] Adds an `AccountState` type with `ToJSON`/`FromJSON` instances, along with serialization tests.
- [x] Adds a `GetAccountState` operation to the Jörmungandr API type, with an associated client implementation.
- [x] Adds integration tests for `GetAccountState` operation. These tests are designed to fail if the wire JSON format for `AccountState` ever changes.

# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

jonathanknowles committed:
Move Jormungandr API types into separate module.

jonathanknowles committed:
Correct typo in comment for module `Jormungandr.Api`.

jonathanknowles committed:
Add structure to public interface of `Jormungandr.Api.Types` module.

jonathanknowles committed:
Remove redundant constraint from function `goldenTestStdTx`.

jonathanknowles committed:
Add type `ApiAccountState` to `Jormungandr.Api.Types`.

jonathanknowles committed:
Add skeleton `GetAccountState` Jormungandr API operation.

jonathanknowles committed:
Provide implementation of `getAccountState`.

jonathanknowles committed:
Reduce unnecessary mixing of `$` and `.` in `mkJormungandrClient`.

jonathanknowles committed:
Add decoding unit tests for example account state JSON objects.

jonathanknowles commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Api/Types.hs:
```diff
@@ -0,0 +1,183 @@
+        ApiAccountState
+            <$> v .: "value"
+            <*> v .:? "delegation"
+            <*> v .: "counter"
```
This should be updated to match this format:
```json
{ "counter": 0
, "delegation":
  { "pools":
    [ [ "5cfbb7f7ec0050499fbd1c79515216c5ad636b97d1944bde8a7db98e7d126803"
      , 1
      ]
    ]
  },
  "value": 1
}
```
Generated with:
```
jcli rest v0 account get ca1skzn99jx8a2rw80gp9uea47tmcndv7gm28vy9as6akevy32204aqwzpsctx --host http://127.0.0.1:43665/api --output-format json
```
See: https://github.com/input-output-hk/cardano-wallet/issues/902#issuecomment-552382037

jonathanknowles committed:
Add integration tests that fetch an `AccountState` via Jörmungandr REST API.

We attempt to fetch:

1. An account that does exist.
2. An account that doesn't exist.

jonathanknowles committed:
Replace `ApiT ChimericAccount` with dedicated `ApiAccountId` type.

jonathanknowles committed:
Add textual encoding tests for the `ApiAccountId` type.

Provide positive, negative, and roundtrip tests.

KtorZ commented on lib/jormungandr/test/unit/Cardano/Wallet/Jormungandr/NetworkSpec.hs:
```diff
@@ -580,7 +597,8 @@ instance Arbitrary S where
+            (+ initialChainLength) <$> choose (0, 2 * fromIntegral k)
         bootstrap <- genNodeOps mockNodeK initialChainLength emptyNode
         let node = applyNodeOps (mconcat bootstrap) emptyNode
         operations <- genNodeOps mockNodeK targetChainLength node
```
I wonder what was the motivation to go over and modify this module which isn't really related to this PR? 
Generally speaking, this kind of changes adds up in a PR and obfuscate a bit the real subject. I am all for refactoring code and updating code but it should always serve the ongoing task: there's no point refactoring a piece of code that is working and that doesn't need new features or bug fixes (a good read on the topic would be: https://martinfowler.com/books/refactoring.html). 

If this kind of stuff is really itchy, I'd recommend separating these in another PR to isolate the rest of the PR from it :+1:

jonathanknowles replied:
@KtorZ wrote:
> I wonder what was the motivation to go over and modify this module which isn't really related to this PR?

The reason is very simple:

1. While working on the `Jormungandr.NetworkSpec` module, I noticed that some of the code didn't follow our team coding standard.
2. It was easy to fix, so I fixed it there and then, placing the changes into a separate commit.

> If this kind of stuff is really itchy, I'd recommend separating these in another PR to isolate the rest of the PR from it +1

It would also be my preference to create a separate PR for such changes.

However, when I did that before, you expressed a preference of placing such changes in a _separate commit_ of the PR to which it was related, rather than a separate PR:

See: https://github.com/input-output-hk/cardano-wallet/pull/117#issuecomment-476240560

Ever since then, I've placed such changes into a _separate commit_, rather than a separate PR. Of course, if you'd prefer that I revert back to putting such changes in a _separate PR_, I'd be happy to do that. Which method would you prefer?

> (a good read on the topic would be: https://martinfowler.com/books/refactoring.html).

Thanks for the recommendation, I read this book many years ago. It's a good book, and I agree with many of the ideas in it. :+1:

KtorZ replied:
@jonathanknowles I am glad you've dug that comment, and glad to see that I am consistent over almost a year about this :sweat_smile: ... The main point of my comment in #117 was actually even in **bold** and it was also the main point of my question here. 

Now looking back at the commit history, I see the diff:

```diff
-    { getTipId = do
+    { getAccountState = \_ -> error "mock getAccountState"
+
+    , getTipId = do
```

I do think it's a bit excessive regarding coding standard and alignment. I mean, this sole module has `17 insertions(+), 37 deletions(-)`, for what is actually a `3 insertions(+), 1 deletion(-)`. And the PR feels also a bit cluttered through code style changes and at this point, it has a negative impact of the "reviewability" of the PR. 

My point is that, these style changes are not helping much the addition (although I understand that a) one can be tempted to adjust lines as one runs into them, b) it may improve one ability to read through the module). 

The right middle-ground IMO is to limit the code style changes to the functions you are actually touching / changing (so, more restrictive than a _module_). This way, there's no need to load the surrounding in mind when reviewing and, we still get to tidy up the codebase steadily.

jonathanknowles replied:
> The right middle-ground IMO is to limit the code style changes to the functions you are actually touching / changing (so, more restrictive than a module).

Fair enough. I've _revised_ the change a bit so that code style changes are restricted to just the functions that are touched by this PR. Let me know if you need it to be reduced further.

**Regarding our PR process in general**:

Perhaps it would be a good idea for us to document our policy on code style fixes, so we have something that we can refer to?

How about having something like the following guidelines:

> 1. A normal PR _may_ include fixes to code style (to meet our team coding standards), but any such fixes should be in a _**single, separate commit**_, and reformatting should be limited to **_only_** the top-level functions that are changed by the PR.

> 2. For _**larger**_ code style fixes that go beyond what's allowed within a normal PR, these should be placed in a _**dedicated**_ PR that introduces no other changes. Such PRs should only be reviewed and merged during _**recovery weeks**_.

Thoughts?

KtorZ replied:
@jonathanknowles I am happy to introduce these additions to the top of the style guide, if others are also happy with it :+1:

KtorZ commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Api/Types.hs:
```diff
@@ -130,12 +157,30 @@ instance Accept Hex where
+        ApiAccountState
+            <$> v .: "value"
+            <*> v .: "delegation"
+            <*> v .: "counter"
```
What about using `Generic` and `genericParseJSON` instead here? This avoid encoding the structure twice (in the type itself, and in the JSON serializer/deserializer) and makes for code that is easier to maintain.

jonathanknowles replied:
@KtorZ wrote:
> What about using Generic and genericParseJSON instead here?

That would also be my preference (normally).

However, the names of the on-the-wire type are somewhat _cryptic_. For example, it's really not obvious that `counter` refers to a total count of the number of _transactions_.

Of course, we _could_ fix this by adding comments to the fields.

But every time we import one of these cryptic names into another module, the reader has to pay the mental cost of translating from the name to the meaning.

I'd argue that it's better to use _self-documenting names_ for the fields, and _pay the translation cost a single time_ in the JSON instances (which we will test).

There's also the issue that the on-the-wire format is subject to change. If the on-the-wire format changes and we use a generic JSON instance, then we have to change the names of the fields in all the modules that import them. But if we use a manually-defined JSON instance above, we can limit the change to just this instance.

Thoughts?

jonathanknowles replied:
@KtorZ I've revised the code to use generic JSON instances for everything, while retaining self-documenting field names on the exported type. See https://github.com/input-output-hk/cardano-wallet/pull/995/commits/06fe68e174c1fd0ec693303da51f39dd14e8af96

KtorZ replied:
> I'd argue that it's better to use self-documenting names for the fields, and pay the translation cost a single time in the JSON instances (which we will test).

I'd slightly disagree and would argue that it is best to stick with Jörmungandr terminology in the http-client for Jörmungandr. The very same way we try to keep the same terminology for the binary formats. This reduces the need for an extra mapping between our names and their names. 

Rather than the field, I'd recommend to make the *types* more expressive (as you did I believe :)) which is what matters the most in the end.

jonathanknowles replied:
Hi @KtorZ

Not sure if you saw my earlier comment, but the current version includes an `ApiAccountState` type with field names that exactly match that the Jörmungandr terminology, with generic JSON instances.

Though since the type is rather nested (and actually requires a pair of types), I've exported a simpler `AccountState` type where much of the nesting has been removed.

This makes client code quite a bit simpler. For example:

```diff
- eitherDecode testAccountState `shouldBe` Right ApiAccountState
-     { currentBalance =
-         ApiT (Quantity testBalance)
-     , totalTransactionCount =
-         ApiT (Quantity testTransactionCount)
-     , delegationInfo = ApiAccountDelegationInfo
-         { stakePools =
-             [ ( ApiT testPoolId1
-               , ApiT (Quantity testPoolRatio1))
-             , ( ApiT testPoolId2
-               , ApiT (Quantity testPoolRatio2))
-             ]
-         }
+ eitherDecode testAccountState `shouldBe` Right AccountState
+     { currentBalance = Quantity testBalance
+     , totalTransactionCount = Quantity testTransactionCount
+     , stakePools =
+         [ (testPoolId1, Quantity testPoolRatio1)
+         , (testPoolId2, Quantity testPoolRatio2)
+         ]
```

See: https://github.com/input-output-hk/cardano-wallet/pull/995/commits/168a09e195e9084ac1980ab8342ec5a29931d326#diff-a778ed4cdf56bfabb154a6af53fdc329L135-L148

Though of course, if you'd prefer a different solution, let me know and I can revise it further.

KtorZ commented on lib/jormungandr/test/unit/Cardano/Wallet/Jormungandr/ApiSpec.hs:
```diff
@@ -11,22 +12,107 @@ module Cardano.Wallet.Jormungandr.ApiSpec
+                              , ApiT (Quantity testPoolRatio2))
+                            ]
+                        }
+                    }
```
What about using `QuickCheck` + `Arbitrary` here to perform those decoding? This is clearly a case where property based testing shines -> generating random test cases trying to find edge-cases, rather than having a possibly non-exhaustive list of unit test. 

Usually, my own preferences is to have a mix of both QC and "standard unit tests". Having simply a one or a couple of hand-written test case like the above to really convince myself that a particular things follow my intuition, and then, hammer it down with QC.

jonathanknowles replied:
@KtorZ wrote:
> What about using `QuickCheck` + `Arbitrary` here to perform those decoding? This is clearly a case where property based testing shines -> generating random test cases trying to find edge-cases, rather than having a possibly non-exhaustive list of unit test.

Fair enough! I've added roundtrip JSON encoding/decoding tests for the `AccountState` type. See: https://github.com/input-output-hk/cardano-wallet/pull/995/commits/08fa510aa1b255bdcbb65bf702f9b866744d06fb

I think it's still worth keeping the handwritten positive unit tests, as these are based on actual examples of JSON output from the `jcli` tool.

KtorZ replied:
> I think it's still worth keeping the handwritten positive unit tests, as these are based on actual examples of JSON output from the jcli tool.

Yes, hence my second point in my original comment. Having a couple of hand-written scenario is useful to build a correct intuition of a particular behavior. But the actual testing is done by QC ^.^

KtorZ commented on lib/jormungandr/test/unit/Cardano/Wallet/Jormungandr/ApiSpec.hs:
```diff
@@ -111,3 +197,20 @@ spec = do
+    "653f9aa1e6c0374528517a37f51d356466cb1b415dcb8d4100772879f41f087c"
+
+mkTestPoolId :: Text -> PoolId
+mkTestPoolId = either (error textError) id . fromText
```
What about calling this `unsafePoolId`, and moving it with other unsafe constructors in `Cardano.Wallet.Unsafe` for better re-usability and consistency ?

jonathanknowles replied:
@KtorZ wrote:
> What about calling this `unsafePoolId`, and moving it with other unsafe constructors in `Cardano.Wallet.Unsafe` for better re-usability and consistency ?

Done: https://github.com/input-output-hk/cardano-wallet/pull/995/commits/1a454beb80b830d8cf6f122d7674fd72fb26ebc0

Anviking replied:
I'd question calling it `unsafePoolId` instead re-using the existing `unsafeDecodeHex`.

Looking at this, I see that the pool id is in hex, but I don't know if `unsafePoolId` or the `FromText` instance of `PoolId` expects that.

I guess this still might be preferable though.

KtorZ commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Api/Types.hs:
```diff
@@ -107,9 +105,22 @@ data ApiStakeDistribution = ApiStakeDistribution
+
+accountIdHumanReadablePart :: Bech32.HumanReadablePart
+accountIdHumanReadablePart =
+    Bech32.unsafeHumanReadablePartFromText "ca"
```
I don't understand this choice of prefix :thinking: ? Is it simply provisional? Although it doesn't matter much for I believe Jörmungandr doesn't pay attention to what prefix is used, I'd suggest using something that is more representative of an account public key :+1:

KtorZ replied:
I just noticed that this is the default prefix used by `jcli` when generating addresses! 
This is unfortunate as this default is no longer in use, but it can be changed with a `--prefix` option !

jonathanknowles replied:
@KtorZ wrote:

> I don't understand this choice of prefix?  Is it simply provisional?

Good question. I had assumed that this was the expected prefix based on the contents of our `genesis.yaml`.

But you're right, Jörmungandr doesn't seem to care what the prefix is. Here's the output of a `jcli` session, when given one of the account addresses in our `genesis.yaml`:

```
$ jcli rest v0 account get ca1skzn99jx8a2rw80gp9uea47tmcndv7gm28vy9as6akevy32204aqwzpsctx --host http://127.0.0.1:46127/api --output-format json | jq -c
{"counter":0,"delegation":{"pools":[["5cfbb7f7ec0050499fbd1c79515216c5ad636b97d1944bde8a7db98e7d126803",1]]},"value":1}
```

Here's the output when given the same account address _re-encoded with a different prefix_ (note that the 6-digit checksum has changed).
```
$ jcli rest v0 account get account1skzn99jx8a2rw80gp9uea47tmcndv7gm28vy9as6akevy32204aqwnfj0ln --host http://127.0.0.1:46127/api --output-format json | jq -c
{"counter":0,"delegation":{"pools":[["5cfbb7f7ec0050499fbd1c79515216c5ad636b97d1944bde8a7db98e7d126803",1]]},"value":1}
```
The result is exactly the same.

Internally, these two `bech32` strings resolve to the same account.

KtorZ replied:
Yes! The human readable part ought to be ignored by softwares for it is really for.. humans :sweat_smile:
`account` sounds like a good choice of prefix in this situation :+1:

jonathanknowles replied:
@KtorZ wrote:

> I'd suggest using something that is more representative of an account public key.

:+1: 

I've changed the prefix to "account". See: https://github.com/input-output-hk/cardano-wallet/pull/995/commits/1639467c242a5f7e769cfb47e28dd6fef32d8a8f

Though of course if this isn't suitable, let me know, and we can change it to something else.

jonathanknowles replied:
Further to the above comment, `AccountId` is now encoded as a hexadecimal string, as this is what the Jormungandr REST API expects. So we no longer need to use a prefix.

See: https://github.com/input-output-hk/cardano-wallet/pull/995#issuecomment-553807074

KtorZ commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Api/Types.hs:
```diff
@@ -73,10 +67,14 @@ import GHC.Generics
 import qualified Servant.API.ContentTypes as Servant
 
+newtype ApiAccountId = ApiAccountId { getApiAccountId :: Bech32.DataPart }
+    deriving (Eq, Show)
```
Having a `DataPart` at this level (in the type representation) is forcing a user-facing encoding onto the users of that API type. One of the main goal of an API client is to abstract concerns such as encoding or protocol details from the client consumer. Since `ApiAccountId` is exposed as a result type, it'd be best to use a raw type that doesn't suggest any particular encoding (`ByteString` or, `ChimericAccount`).

jonathanknowles replied:
@KtorZ wrote:
> Since `ApiAccountId` is exposed as a result type, it'd be best to use a raw type that doesn't suggest any particular encoding.

Fair enough! I've updated `AccountId` so that the underlying representation is a `ByteString`, converting to Bech32 only within the `ToText` instance, using a customary human-readable prefix of "account" (which we can change if need to).

See: https://github.com/input-output-hk/cardano-wallet/pull/995/commits/1639467c242a5f7e769cfb47e28dd6fef32d8a8f

jonathanknowles replied:
Further to the above comment, `AccountId` is now encoded as a hexadecimal string, as this is what the Jormungandr REST API expects. See: https://github.com/input-output-hk/cardano-wallet/pull/995#issuecomment-553807074

jonathanknowles committed:
Use generic JSON encoding/decoding instead of hand-written instances.

Also, export the `AccountState` type rather than the `ApiAccountState` type.

This allows client code to be simpler.

jonathanknowles committed:
Add `ToJSON` instance for `AccountState` type.

jonathanknowles committed:
Add roundtrip JSON encoding tests for `AccountState` type.

jonathanknowles committed:
Rename `mkTestPoolId` to `unsafePoolId`.

Additionally, move it to `Cardano.Wallet.Unsafe`.

jonathanknowles committed:
Rename `ApiAccountId` to `AccountId`.

jonathanknowles committed:
Use `ByteString` as the underlying representation of `AccountId`.

Additionally, use hex-encoding for the textual representation of `AccountId`.

This is consistent with what the Jörmungandr REST API expects.

jonathanknowles committed:
Adjust code to comply with team coding standard.

https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#limit-line-length-to-80-characters

Anviking commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Api/Client.hs:
```diff
@@ -158,7 +158,7 @@ mkJormungandrClient mgr baseUrl = JormungandrClient
         let action = cGetAccountState (ApiT accountId)
         run action >>= \case
             Left (FailureResponse e) | responseStatusCode e == status404 ->
-                return . Left . ErrGetAccountStateAccountNotFound $ accountId
```
I kind of like the `. . . $` style. You don't?

jonathanknowles replied:
> I kind of like the `. . . $` style. You don't?

On the contrary, I also like that style when the part to the right of the `$` is complex.

But in this particular example, I was looking at a choice between:
* `return . Left . ErrGetAccountStateAccountNotFound $ accountId`
* `return $ Left $ ErrGetAccountStateAccountNotFound accountId`
The second one seemed simpler, as it has one fewer operator, so I went for that.

But I don't have a strong feeling either way :)

Anviking commented on lib/jormungandr/test/unit/Cardano/Wallet/Jormungandr/ApiSpec.hs:
```diff
@@ -11,22 +12,107 @@ module Cardano.Wallet.Jormungandr.ApiSpec
+            it "With 0 stake pools" $ do
+
+                let testAccountState = Aeson.encode [aesonQQ|
+                        { "value": #{testBalance}
```
Interesting!

I guess these tests are good to have.

jonathanknowles replied:
Yes I agree it's nice to have these! I took the output from `jcli` and modified it minimally to make these tests.

Of course, we also have the JSON round trip property tests to give us more assurance.

Anviking commented on lib/jormungandr/test/integration/Cardano/Wallet/Jormungandr/NetworkSpec.hs:
```diff
@@ -140,6 +154,23 @@ spec :: Spec
+                    , totalTransactionCount =
+                        ApiT (Quantity 0)
+                    , delegationInfo = ApiAccountDelegationInfo
+                        { stakePools = [ (testPoolId, ApiT $ Quantity 1) ]
```
👍

Anviking commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Api/Types.hs:
```diff
@@ -75,27 +78,22 @@ import qualified Servant.API.ContentTypes as Servant
 
+data AccountState = AccountState
+    { currentBalance
+        :: !(Quantity "lovelace" Word64)
```
😅  Is this really needed?

If I understand things correctly
1. `AccountState` is still an "Api type" with only internal use by the `Jormungandr.Client`
2. We now have two of those types: `AccountState` and `ApiAccountState`, with the only difference being the field names.

In general it sounds better to me to go with the existing names/concepts and add comments explaining them if necessary.

**edit**: seems `AccountState` is used in the `getAccountState` signature.

jonathanknowles replied:
> AccountState and ApiAccountState, with the only difference being the field names.

Actually, `ApiAccountState` is a more complicated type, because:

1. It requires an additional type (`ApiAccountDelegationState`) to encode the list of delegations. This is because our `Api...` types generally have a one-to-one mapping with the JSON on-the-wire encoding, which enables us to use a generic JSON encoding, and the JSON encoding is also nested.

2. It nests many things in `ApiT`. But for consumers of the type, `ApiT` is just something to unwrap. It has no real purpose outside of the `Api.Types` module.

Have a look at this diff, which shows a piece of client code, when the return type of `getAccountState` was changed from `ApiAccountState` to `AccountState`:

```diff
- eitherDecode testAccountState `shouldBe` Right ApiAccountState
-     { currentBalance =
-         ApiT (Quantity testBalance)
-     , totalTransactionCount =
-         ApiT (Quantity testTransactionCount)
-     , delegationInfo = ApiAccountDelegationInfo
-         { stakePools =
-             [ ( ApiT testPoolId1
-               , ApiT (Quantity testPoolRatio1))
-             , ( ApiT testPoolId2
-               , ApiT (Quantity testPoolRatio2))
-             ]
-         }
+ eitherDecode testAccountState `shouldBe` Right AccountState
+     { currentBalance = Quantity testBalance
+     , totalTransactionCount = Quantity testTransactionCount
+     , stakePools =
+         [ (testPoolId1, Quantity testPoolRatio1)
+         , (testPoolId2, Quantity testPoolRatio2)
+         ]
```

Anviking commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Api/Types.hs:
```diff
@@ -80,7 +80,7 @@ import qualified Data.Aeson.Types as Aeson
 import qualified Servant.API.ContentTypes as Servant
 
-newtype ApiAccountId = ApiAccountId { getApiAccountId :: Bech32.DataPart }
+newtype AccountId = AccountId { getAccountId :: Bech32.DataPart }
```
Similarly to a previous comment: `AccountId` seems to be escaping the `Jormungandr.Client` module.

It is used here:

```haskell
    { getAccountState
        :: AccountId
        -> ExceptT ErrGetAccountState m AccountState
``` 

My perception was that `Api.Types` wasn't supposed to "escape" like that.

An alternative would be to put this in a `Jormungandr.Primitive.Types` module, but tbh, I wonder if this isn't neater as you've done it.

jonathanknowles replied:
> An alternative would be to put this in a Jormungandr.Primitive.Types module, but tbh, I wonder if this isn't neater as you've done it.

That's a good point. We don't currently have a `Jormungandr.Primitive.Types` module. Perhaps we can create one in a future PR?

Anviking commented on lib/jormungandr/test/unit/Cardano/Wallet/Jormungandr/ApiSpec.hs:
```diff
@@ -11,22 +12,107 @@ module Cardano.Wallet.Jormungandr.ApiSpec
+                        , "counter": #{testTransactionCount}
+                        , "delegation":
+                            { "pools":
+                                [ [#{toText testPoolId1}, #{testPoolRatio1}] ]
```
I wonder if the test could be more useful without the `toText` here.

What if we explicitly typed the hex in the json, and the explicitly typed the raw byte string as part of the `ApiAccountState`? Or using `unsafeFromHex`/`unsafeToHex` helpers.

I think the hex encoding part is something that can easily be missed, and it would be good if our tests would guard against it.

I'm also not sure what role this test plays in comparison with the aeson golden tests.

jonathanknowles replied:
> I wonder if the test could be more useful without the toText here.

I think we do want to test that the input (JSON) and output (decoded value) share _the same_ test pool ID, hence using a common identifier here is one way to check.

Perhaps I have not understood your point correctly though. Could you code up an example of what you mean?

Anviking reviewed (approved):
Some thoughts / open questions, but looks good to me!

Also, I would have tried squashing a few commits (like 15259830c4ed555d37783f5970e9320b079d949f) but the commit history was helpful and easy to go through 🙏 

I will leave bors-ing to you or someone else.
